File: packages/backstage-ts/test/script-registry.test.ts
/**
 * ScriptRegistry tests
 */
import { describe, test, expect, beforeAll, afterAll } from 'bun:test';
import { ScriptRegistry, type ScriptDef } from '../src/script-registry';
import { type RedisClient } from '../src/types';

let redis: RedisClient;
let registry: ScriptRegistry;

beforeAll(async () => {
  redis = new Bun.RedisClient('redis://localhost:6379');
  registry = new ScriptRegistry(redis);
});

afterAll(async () => {
  const keys = await redis.send('KEYS', ['backstage:script:*']);
  if (keys && Array.isArray(keys) && keys.length > 0) {
    await redis.send('DEL', keys as string[]);
  }
});

describe('ScriptRegistry', () => {
  test('loads script correctly', async () => {
    const scriptName = 'test-load';
    await registry.load({
      [scriptName]: {
        script: 'return ARGV[1]',
        keys: {},
      },
    });

    expect(registry.has(scriptName)).toBe(true);
    const script = registry.get(scriptName);
    expect(script).toBeTruthy();
    expect(script?.sha).toBeTruthy();
  });

  test('executes script with arguments', async () => {
    const scriptName = 'test-args';
    await registry.load({
      [scriptName]: {
        script: 'return {KEYS[1], ARGV[1]}',
        keys: { testKey: 1 },
      },
    });

    const result = await registry.run<[string, string]>(
      scriptName,
      { testKey: 'my-key' },
      ['my-arg'],
    );

    expect(result).toHaveLength(2);
    expect(result[0]).toBe('my-key');
    expect(result[1]).toBe('my-arg');
  });

  test('handles NOSCRIPT error automatically', async () => {
    const scriptName = 'test-noscript';
    await registry.load({
      [scriptName]: {
        script: 'return "recovered"',
        keys: {},
      },
    });

    // Manually flush scripts from Redis to simulate restart/cache eviction
    await redis.send('SCRIPT', ['FLUSH']);

    // Registry still has the SHA, but Redis doesn't
    // expected behavior: execute -> fail with NOSCRIPT -> reload -> retry -> success

    const result = await registry.run<string>(scriptName, {}, []);
    expect(result).toBe('recovered');

    // Check that SHA might have changed or at least is valid again
    const script = registry.get(scriptName);
    expect(script?.sha).toBeTruthy();
  });

  test('throws on missing keys', async () => {
    const scriptName = 'test-missing-keys';
    await registry.load({
      [scriptName]: {
        script: 'return KEYS[1]',
        keys: { requiredKey: 1 },
      },
    });

    const promise = registry.run(scriptName, {}, []);
    expect(promise).rejects.toThrow("Missing required key 'requiredKey'");
  });

  test('throws on unregistered script', async () => {
    const promise = registry.run('non-existent', {}, []);
    expect(promise).rejects.toThrow("Script 'non-existent' is not registered");
  });

  test('validates key indices', async () => {
    const scriptName = 'test-bad-indices';
    await registry.load({
      [scriptName]: {
        script: 'return "ok"',
        keys: {
          key1: 1,
          key3: 3, // Skipping 2
        },
      },
    });

    const promise = registry.run(scriptName, { key1: 'k1', key3: 'k3' }, []);
    expect(promise).rejects.toThrow('Missing key for index 2');
  });

  test('Lua execution error propagates', async () => {
    const scriptName = 'test-lua-error';
    await registry.load({
      [scriptName]: {
        script: 'return redis.call("INVALID_COMMAND")',
        keys: {},
      },
    });

    const promise = registry.run(scriptName, {}, []);
    expect(promise).rejects.toThrow();
  });
});

File: packages/backstage-go/script_registry_test.go
package backstage

import (
	"context"
	"testing"
	"strings"

	"github.com/redis/go-redis/v9"
)

func TestScriptRegistry(t *testing.T) {
	ctx := context.Background()

	// Setup Redis client
	rdb := redis.NewClient(&redis.Options{
		Addr: "localhost:6379",
	})
	defer rdb.Close()

	// Verify Redis connection
	if err := rdb.Ping(ctx).Err(); err != nil {
		t.Skipf("Skipping test, redis unavailble: %v", err)
	}

	// Clean up scripts
	defer rdb.ScriptFlush(ctx)

	t.Run("LoadsScriptCorrectly", func(t *testing.T) {
		registry := NewScriptRegistry(rdb)
		scriptName := "test-load"
		
		err := registry.Load(ctx, map[string]ScriptDef{
			scriptName: {
				Script: "return ARGV[1]",
				Keys:   map[string]int{},
			},
		})
		if err != nil {
			t.Fatalf("Load failed: %v", err)
		}

		if !registry.Has(scriptName) {
			t.Errorf("Expected registry to have script %q", scriptName)
		}
		
		if sha := registry.GetSHA(scriptName); sha == "" {
			t.Errorf("Expected SHA for script %q", scriptName)
		}
	})

	t.Run("ExecutesScriptWithArguments", func(t *testing.T) {
		registry := NewScriptRegistry(rdb)
		scriptName := "test-args"

		err := registry.Load(ctx, map[string]ScriptDef{
			scriptName: {
				Script: "return {KEYS[1], ARGV[1]}",
				Keys:   map[string]int{"testKey": 1},
			},
		})
		if err != nil {
			t.Fatalf("Load failed: %v", err)
		}

		result, err := registry.Run(ctx, scriptName, map[string]string{"testKey": "my-key"}, "my-arg")
		if err != nil {
			t.Fatalf("Run failed: %v", err)
		}

		response, ok := result.([]interface{})
		if !ok || len(response) != 2 {
			t.Fatalf("Unexpected response format: %v", result)
		}

		if response[0] != "my-key" {
			t.Errorf("Expected result[0] to be 'my-key', got %v", response[0])
		}
		if response[1] != "my-arg" {
			t.Errorf("Expected result[1] to be 'my-arg', got %v", response[1])
		}
	})

	t.Run("HandlesNOSCRIPTErrorAutomatically", func(t *testing.T) {
		registry := NewScriptRegistry(rdb)
		scriptName := "test-noscript"

		err := registry.Load(ctx, map[string]ScriptDef{
			scriptName: {
				Script: "return 'recovered'",
				Keys:   map[string]int{},
			},
		})
		if err != nil {
			t.Fatalf("Load failed: %v", err)
		}

		// Flush scripts to simulate restart/eviction
		if err := rdb.ScriptFlush(ctx).Err(); err != nil {
			t.Fatalf("ScriptFlush failed: %v", err)
		}

		// Should recover automatically
		result, err := registry.Run(ctx, scriptName, map[string]string{})
		if err != nil {
			t.Fatalf("Run failed after flush: %v", err)
		}

		if result != "recovered" {
			t.Errorf("Expected result 'recovered', got %v", result)
		}

		if sha := registry.GetSHA(scriptName); sha == "" {
			t.Errorf("Expected SHA to be present after recovery")
		}
	})

	t.Run("ThrowsOnMissingKeys", func(t *testing.T) {
		registry := NewScriptRegistry(rdb)
		scriptName := "test-missing-keys"

		err := registry.Load(ctx, map[string]ScriptDef{
			scriptName: {
				Script: "return KEYS[1]",
				Keys:   map[string]int{"requiredKey": 1},
			},
		})
		if err != nil {
			t.Fatalf("Load failed: %v", err)
		}

		_, err = registry.Run(ctx, scriptName, map[string]string{})
		if err == nil {
			t.Error("Expected error for missing key, got nil")
		} else {
			if !strings.Contains(err.Error(), "missing required key") {
				t.Errorf("Expected error message to contain 'missing required key', got: %v", err)
			}
		}
	})

	t.Run("ThrowsOnUnregisteredScript", func(t *testing.T) {
		registry := NewScriptRegistry(rdb)
		_, err := registry.Run(ctx, "non-existent", nil)
		if err == nil {
			t.Error("Expected error for unregistered script, got nil")
		}
	})

	t.Run("ValidatesKeyIndices", func(t *testing.T) {
		registry := NewScriptRegistry(rdb)
		scriptName := "test-bad-indices"

		err := registry.Load(ctx, map[string]ScriptDef{
			scriptName: {
				Script: "return 'ok'",
				Keys: map[string]int{
					"key1": 1,
					"key3": 3,
				},
			},
		})
		if err != nil {
			t.Fatalf("Load failed: %v", err)
		}

		// This will technically fail inside Run because of verification logic
		// But let's check if the gap detection works
		_, err = registry.Run(ctx, scriptName, map[string]string{
			"key1": "k1",
			"key3": "k3",
		})
		if err == nil {
			t.Error("Expected error for non-contiguous indices, got nil")
		} else {
			if !strings.Contains(err.Error(), "invalid key index") {
				t.Errorf("Expected error message about invalid key index, got: %v", err)
			}
		}
	})
}


